Class-based transactions

GustavoCaso opened the pull request:
Implement class-based transactions based on design outlined in #52. 

Resolves #52.

<hr>

_Original PR notice:_

This implements this #52 

This still experimental, but before adding more code and more logic to the library I wanted to share @timriley the simplest example so we can discuss how to continue working on this.

All feedback is gratefully received 😄 

Thanks

GustavoCaso committed:
WIP

GustavoCaso committed:
return Transaction::DSL to original state

GustavoCaso committed:
Clean module builder

GustavoCaso committed:
Allow pass specific step operations at initialize

GustavoCaso committed:
method name

GustavoCaso committed:
make test pass

GustavoCaso commented:
@timriley so far with this implementation we can replacement dependencies to make testing easier:

 ```ruby
# Provide an explicit operation for step one
MyTransaction.new(one: different_operation)
```

But I think this approach is not going to help because one key feature would be been able to wrap existing steps. I was thinking maybe we could instead of creating the step within the macro to store them in some class level variables like `step_info` that store step type `map, raw, tee` and the name of the step, with that in place at initialization time we could dynamically generate the instance method with the step names, that way the user can wrap them. 

Also since we want that the user can provide extra locals step, I was thinking of modifying the `Transaction` class to instead of getting `steps` and `matcher` it can get the instance of the class so we can create the steps checking for them in the container or in the class.

What do you think?

timriley committed:
WIP

timriley commented:
Hey @GustavoCaso, thanks for pushing this forward! I've spent some time this morning sketching out in code the kind of direction we probably wanna take this, but I feel I haven't got to a good spot either. You can see I've pushed up a branch here (https://github.com/dry-rb/dry-transaction/compare/class_base_transaction...class_based_transactions_timriley) - feel free to take a look and let me know if you have any thoughts so far. I feel I need to experiment just a tad more before I can summarise things nearly in words for you. Roughly, though:

- I've made this work in a way where we don't _require_ container objects as operations, since users can just provide local methods for some/all of their steps, if they want
- I'm not happy with where the container object resolution is, though. I feel like including (or prepending?) an `#initialize` might be good, which is the only place we try and resolve stuff, and then we can pass them through to the standard initialise for storing those objects in local vars

Sorry I can't get back to you with anything better so far, let's just keep the conversation going.

GustavoCaso committed:
Add more tests cases

GustavoCaso committed:
Fix integration tests

GustavoCaso committed:
Add step args methods to pass extra arguments to steps

GustavoCaso committed:
Remove unwanted tests and encapsulate Builder

GustavoCaso commented:
@timriley I added the method for passing extra step arguments, and update all integration test to use new class base transaction.

Also, remove all not used code to make things easier while reviewing this PR.

When we are to place we are happy to push this change, I can start with the documentation both in the project and in the `dry-rb.org` website.

timriley committed:
Remove support for steps with operation procs passed to with:

Now that steps map to instance methods, we don’t need this feature anymore.

timriley committed:
Add a FIXME note

timriley committed:
Improve container integration and tighten up overall behavior 

- Move resolution of operations from container into `#initialize` in a prepended resolver module. Deferring the resolution of objects from the container until the transaction is instantiated means we won’t encounter errors if the container isn’t fully populated at the time when the class is defined.
- Only try and resolve operations from the container if a `with:` option is supplied to the step. Now that a container is not required to work with dry-transaction, I don’t think that it makes sense to try and resolve operations from the container using the step’s local name, which should be meaningful within the current transaction class only.
- Move the asserts for step arity and valid step names into more appropriate places
- When “currying” a transaction with extra step args, create a copy of the transaction rather than mutating it
- When instantiating a transaction, build a local copy of all the steps with the appropriate operations passed to each (local methods if defined or the resolved container operations)
- Various minor renames

timriley committed:
Move InstanceMethods module into its own source file

timriley committed:
Switch Dry::Transaction from a class to a module

We’re never creating or returning Dry::Transaction instances, so this doesn’t need to be a class

GustavoCaso committed:
Allow transaction definition to be define without `with` keyword

GustavoCaso committed:
change name method from apply to curry for Step instances

GustavoCaso committed:
Guard clause if no argument is provided, add test for `curry` method

timriley commented on lib/dry/transaction/step.rb:
```diff
@@ -27,12 +29,24 @@ def initialize(step_adapter, step_name, operation_name, operation, options, call
 
-      def with_call_args(*call_args)
-        self.class.new(step_adapter, step_name, operation_name, operation, options, call_args, &block)
+      def curry(operation: UNDEFINED, call_args: UNDEFINED)
```
I think my TODO about renaming this to `#curry` was actually out of date. I don't think curry as a name makes sense if we're passing more than just call args. Would you mind renaming again to `#with`? That's more consistent with similar methods elsewhere in dry-rb land.

timriley commented:
This is looking pretty good, now! Here's what I think we have left:

- Merge all the stuff from `spec/integration/class_based_spec.rb` into `spec/integration/transaction_spec.rb` (since _every_ transaction is class-based now, I don't see the need to have a special spec file for this)
- Consider if we can bring back `#prepend`/`#append`/`#insert`/`#remove`. I have a feeling that some people will be using these, and if there's a way we can still support them, it'd be nice.

Once we've done all this, I might ping a few more people for review since this is a big change to how dry-transaction works.

I'll take a look at the TravisCI build stuff in a day or too, btw.

GustavoCaso committed:
rename `curry` to `with`

GustavoCaso committed:
Merge test into one file

GustavoCaso committed:
Add back append, prepend, remove and insert

timriley committed:
Revert "Add back append, prepend, remove and insert"

This reverts commit 7049a880a5a0a90a4ae5f19edca8043f22ffc322.

timriley committed:
Remove FIXME

timriley committed:
Remove spurious private

timriley committed:
Make require more direct

timriley committed:
Avoid prepending an anonymous module from OperationResolver

timriley committed:
Remove customisable matcher

Now that we’re class based, users can provide their own matchers from their own superclasses/modules

timriley committed:
Define class DSL from within its own module subclass

timriley commented:
@GustavoCaso Made some more changes here. I decided that the append/prepend/insert stuff was too fraught with potential issues for us to keep it (especially when we're potentially mixing different transaction classes with their own different configurations/internal state/other private helper methods, etc.).

timriley committed:
Stop supporting blocks for steps

Now that we can wrap them in local methods, we shouldn’t need this

timriley committed:
Allow plain Dry::Transaction module to be mixed in

This is what you’d be doing if you were building a transaction without relying a container or specifying any other options. In this situation, `include Dry::Transaction` is much more natural than `include Dry::Transaction()`.

timriley committed:
Use double quotes

timriley committed:
Filter out spec/ dir for SimpleCov

timriley committed:
Update RubyGems and bundler on CI

timriley committed:
Bump bundler dependency

timriley committed:
Update Ruby versions for CI

timriley committed:
Tidy up transaction specs

timriley committed:
Add Dry::Transaction::Operation convenience module

timriley committed:
Remove .bundle/config

timriley committed:
Use step name for finding injected step operation

timriley committed:
Include Dry::Monads::Either::Mixin when including Transaction module

This makes it possible to use the shorthand `Right` and `Left` methods in steps defined by local instance methods.

timriley committed:
Update API docs

timriley commented:
@GustavoCaso We did it! 🎉 

Merging this now and will hopefully release tomorrow along with some updated docs over in https://github.com/dry-rb/dry-rb.org/pull/180.
